src/server/replication_messages.rs

```diff
@@ -0,0 +1,264 @@
+    pub(super) fn register_entity(&mut self) {
+        let data_size = self.buffer.as_slice().len() - self.buffer.entity_data_pos() as usize;
+        self.entities.push((self.buffer.data_entity(), data_size));
+    }
+
+    /// Returns `true` is message contains any written data.
+    fn is_sendable(&self) -> bool {
+        !self.buffer.as_slice().is_empty()
+    }
+
+    /// Splits message according to entities inside it and sends it to the specified client.
+    ///
+    /// Does nothing if there is no data to send.
+    fn send(
+        &mut self,
+        server: &mut RenetServer,
+        client_info: &mut ClientInfo,
+        last_change_tick: LastChangeTick,
+        replicon_tick: RepliconTick,
+        tick: Tick,
+    ) -> bincode::Result<()> {
+        if !self.is_sendable() {
+            trace!("no updates to send for client {}", client_info.id);
+            return Ok(());
+        }
+
+        trace!("sending update message(s) to client {}", client_info.id);
+        const TICKS_SIZE: usize = 2 * mem::size_of::<RepliconTick>();
+        let mut header = [0; TICKS_SIZE + mem::size_of::<u16>()];
+        bincode::serialize_into(&mut header[..], &(*last_change_tick, replicon_tick))?;
+
+        let mut slice = self.buffer.as_slice();
+        let mut entities = Vec::new();
+        let mut message_size = 0;
+        for &(entity, data_size) in &self.entities {
+            const MAX_PACKET_SIZE: usize = 1200; // https://github.com/lucaspoffo/renet/blob/acee8b470e34c70d35700d96c00fb233d9cf6919/renet/src/packet.rs#L7
+
+            // Pack small messages ordered after large messages into the large message's packet ('pack back' strategy).
+            if message_size == 0 || ((message_size + header.len()) % MAX_PACKET_SIZE) < data_size {
+                entities.push(entity);
+                message_size += data_size;
+            } else {
+                let (message, remaining) = slice.split_at(message_size);
+                slice = remaining;
+                message_size = data_size;
+
+                let update_index = client_info.register_update(tick, entities.clone());
+                bincode::serialize_into(&mut header[TICKS_SIZE..], &update_index)?;
+
+                server.send_message(
+                    client_info.id,
+                    ReplicationChannel::Unreliable,
+                    Bytes::from_iter(header.into_iter().chain(message.iter().copied())),
+                );
+
+                entities.clear();
+            }
+        }
+
+        if !slice.is_empty() {
+            let update_index = client_info.register_update(tick, entities);
+            bincode::serialize_into(&mut header[TICKS_SIZE..], &update_index)?;
+
+            server.send_message(
+                client_info.id,
+                ReplicationChannel::Unreliable,
+                Bytes::from_iter(header.into_iter().chain(slice.iter().copied())),
```

Comment:
Slices and arrays implement [ExactSizeIterator](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html) and chaining will also result into iterator that implements it.

But no, it's not efficient because every `Bytes` creating is an allocation. Unfortunately, I can't avoid it because `Bytes`, while cheaply clonable, can't grow in size.
It would be much better if renet accepted anything with `IntoIter<Item = u8>` and copied data from them into its message pools. But it's something to discuss with renet's author.
`@RJ` did it nicely for https://github.com/RJ/pickleback (it accepts slices right now, but the API can be easily extended to iterators).